Add syntax support for PARTITIONED BY/CLUSTERED BY in INSERT queries

LakshSingla opened the pull request:
**NOTE: This PR and the description was originally written while keeping PARTITION BY and CLUSTER BY clauses in mind. The changes might not be reflected in the description and the comments.**

This PR aims to add parser changes for supporting `PARTITIONED BY` and `CLUSTERED BY` as proposed in the issue https://github.com/apache/druid/issues/11929. 

### Description
This PR adds the ability to execute the queries of the form
```
INSERT INTO druid.dst SELECT * FROM druid.foo PARTITION BY 'day' CLUSTER BY 1, 2
```
Since `PARTITION BY` and `CLUSTER BY` are non-standard SQL keywords whose support is not provided in Apache Calcite (which is used for parsing the SQL queries), some changes are made into the `pom.xml` of `druid-sql` to accommodate for the changes made to the parser. 

### Proposal
Calcite is a general-purpose library, and while it does support parsing of SQL that should cover most of the dialects of SQL, one can also provide one's own SQL parser to it, to parse custom SQL statements. 

#### Pre Read
(This is my current understanding and might have some flaws in it, please feel free to point out where I am wrong)

**Provision in Calcite for extending the grammar**
Calcite provides the option of setting the parser factory in `SqlParser.Config` which could be anything that implements `SqlParserImplFactory`. Therefore as far as Calcite is concerned one can rewrite the whole SQL parser from scratch in Java and can pass it to the SqlParser.Config (which is not ideal and tedious). Following is the minimalistic code snippet which achieves the same:

```
  static final SqlParser.Config PARSER_CONFIG = SqlParser
      .configBuilder()
      .setParserFactory(new CustomImplementationExtendingSqlParserImplFactory()) // Custom sql parser factory
      .build();
```

For providing the default parser, Calcite uses JavaCC, which is a parser generator (`Parser.jj`) to specify the grammar in [EBNF](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form), which gets compiled to a Java class. Ideally, we would like to add or extend our custom grammar production rules in this `Parser.jj`. Luckily, Calcite does support extending this Parser.jj via the use of Apache FreeMarker (FTL) templates. 

Therefore, in a nutshell, we would like to recompile the default parser by providing our custom rules (through template), and then pass it to the `SqlParser.Config`.

#### Requirements
1. We don’t want to copy-paste Calcite’s `Parser.jj` manually and modify it directly. This is because a lot of changes do happen in Calcite's main branch, and we would ideally want to keep in line with those changes without having to copy and merge the `Parser.jj` from upstream every time. Therefore we want to reuse templates as much as possible. 
2. Automate the whole process of templating and compiling, so that it gets incorporated into the build process.
3. Original SQL functionality shouldn't be affected. 

#### Modifications to the POM
The stages added to the build process look a lot similar to the one used by the project: [dask-sql](https://github.com/dask-contrib/dask-sql) which also extends the grammar for its own purpose. Therefore, we can use other "druidesque" methods/plugins to achieve the following, but the overall steps remain the same:
1. Copy the resources to `target/codegen` directory.
    a. Copy the default `Parser.jj` from Calcite’s jar, to the build directory
    b. Copy our FTL files, which contains our custom rules, and `config.fmpp` files, which contain the information to be plugged in to the parser.
2. Generate consolidated `Parser.jj` which contains the default Calcite's Parser.jj along with our custom rules, using fmpp plugin
3. With the above file as input, and using JavaCC plugin create the custom parser implementation `DruidSqlParserImpl.java`. This implementation would be passed to the SqlConfig for parsing the SQL strings.

#### Production Grammar added
```
/*
customInsert() -> insert() [PARTITION BY stringLiteral()] [CLUSTER BY clusterItems()]
clusterItems() -> (similar to what can be added in front of an ORDER BY clause in a SELECT statement)
*/
SqlNode DruidSqlInsert() :
{
    insertNode = SqlInsert()
    [
      <PARTITION> <BY>
      partitionBy = StringLiteral()
    ]
    [
      <CLUSTER> <BY>
      clusterBy = ClusterItems()
    ]
    {
  if(partitionBy == null && clusterBy == null) {
    return insertNode;
  }
  if(!(insertNode instanceof SqlInsert)) {
    return insertNode;
  }
  SqlInsert sqlInsert = (SqlInsert) insertNode;
  return new DruidSqlInsert(sqlInsert, partitionBy, clusterBy);
}
}
```

#### Class Design
For the custom insert statement, we are creating a class `DruidSqlInsert`, which inherits from the `SqlInsert`. The custom class contains the extra information (like the clause attached to the PartitionBy) while passing the information from the main INSERT node to the parent.
Therefore from the custom class, we can extract the extra information as and when required, while passing the same object down for the further conversion to Rel/DruidRel objects. (NOTE: Those objects however won't have any knowledge of these extra clauses added, that would be manually plugged). 


#### Integration with the DruidPlanner
1. Once the `String` gets parsed as a SqlNode, we get an instance of `DruidSqlInsert` (if the statement is an INSERT statement with either of the PARTITION BY or CLUSTER BY clauses. 
2. After performing some validation, (e.g. SELECT query inside the INSERT (with a CLUSTER BY) should not have an ORDER BY clause, we extract the information passed to the CLUSTER BY.
3. In the SELECT query, we add the information passed to the CLUSTER BY, as its ORDER BY. 

#### Caveats to the Design
(Most of the following arise due to how we are extending Apache Calcite's grammar, and the limited number of regions we can actually make the changes).

1. The biggest caveat to the approach is that, in the Calcite's grammar: [LINK](https://calcite.apache.org/docs/reference.html)  we are NOT extending the `INSERT` production rule. Instead, we are extending the top-level `statement` production rule to look like:
```
statement:
      customInsertStatement
  |   resetStatement
  ....
  |   insert
  |   update
```
This is because in the original Parser.jj, there is a limited number of places wherein we can insert our custom statements. 
The direct implication of this is that `EXPLAIN PLAN FOR SELECT * FROM druid.foo CLUSTER BY dim1` doesn't work, since there is no `customExplainStatement` or its equivalent. To make this work, we would have to add our own custom EXPLAIN rule as well. 

2. As a direct consequence of the above, `customInsertStatement` actually overshadows the `insert` statement. That is any statement that begins with `INSERT` is going to be interpreted as `customInsertStatement`, irrespective of whether it uses custom clauses or not. Therefore, we have to account for simple `insert` statements with our custom rule. 

3. There is no support for our custom syntax at the `RelNode` level. Conversion from SQL string to Druid Native queries look like this: `String` -> `SqlNode` -> `RelNode` -> `DruidRel`. The custom syntax is only present till the conversion from String to SqlNode. After that, the custom parameters are extracted in the `ParsedNodes` present in `DruidPlanner`, and not sent further down to the RelNode/DruidRel's. This is because there is no rule for parsing the custom nodes in the `SqlToRelConverter.class`. This possibly implies that once the native query is created, parameters passed in say `PARTITION BY` would need to be manually injected before executing the query. 

4. The `CLUSTER BY` and `PARTITION BY` clauses are linked to the `INSERT` statement, as opposed to the `SELECT` statements used inside them. Therefore the query looks like `INSERT INTO dst (SELECT * FROM foo) PARTITION BY 'day' CLUSTER BY dim1` rather than `INSERT INTO dst (SELECT * FROM foo PARTITION BY 'day' CLUSTER BY dim1)`. 

#### Alternatives design ideas
1. The most 'ideal' however unmaintainable way to achieve the same would be to copy the Parser.jj from Calcite project directly, and make the changes there. This would allow us to directly make the changes in the SqlInsert's production rule (and mitigate caveats 1 & 2). However, doing so is not [recommended](https://calcite.apache.org/docs/adapter.html#extending-the-parser). 
2. Instead of `DruidSqlInsert` inheriting from the `SqlInsert`, it can not do so. To make the query work, we would need to replace the query node (in ParsedNodes). The current class design ensures that no replacement of nodes is required. 

#### Note
1. In the newer version of Calcite, there is a `default_config.fmpp` which can be used to supply the boilerplate code in `config.fmpp` (https://github.com/apache/calcite/pull/2167). On migrating to newer version, that file can be used to reduce the code in `config.fmpp` significantly. 

PS: Currently the build process is failing due to spotbugs/forbiddenapis being run on the generated parser files, and due to failing `mvn dependency:analyze` check. 

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

<hr>

##### Key changed/added classes in this PR

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

LakshSingla commented:
LGTM failed due to `forbiddenapis` failing on the generated parser file.

abhishekagarwal87 commented on sql/src/main/codegen/includes/insert.ftl:
```diff
@@ -0,0 +1,65 @@
+    insertNode = SqlInsert()
+    [
+      <PARTITION> <BY>
+      partitionBy = StringLiteral()
```
this would mean that we have to do `partition by 'floor(__time to day)'`. Instead, it would be nice to do `partition by floor(__time to day)`. so what comes after partitionBy should be captured as an expression instead.

abhishekagarwal87 replied:
never mind. we accept a literal in the ingestion spec itself. We are good here.

cryptoe commented on sql/pom.xml:
```diff
@@ -180,6 +180,11 @@
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>1.7.25</version>
```
Can this come from the parent pom ?

LakshSingla replied:
Hi, I just remembered why I didn't use the version from the parent pom. Their is a parent class variable `<slf4j.version>1.7.12</slf4j.version>`, however the sl4j-api is using a different version in the parent pom, which is also hardcoded.

cryptoe commented on sql/pom.xml:
```diff
@@ -255,6 +260,140 @@
+      <plugin>
+        <groupId>com.googlecode.fmpp-maven-plugin</groupId>
+        <artifactId>fmpp-maven-plugin</artifactId>
+        <version>1.0</version>
```
Push to parent pom would be helpful

LakshSingla replied:
PTAL at the poms once again, updated with the relevant changes.

cryptoe commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,433 @@
+
+    # List of new keywords. Example: "DATABASES", "TABLES". If the keyword is not a reserved
+    # keyword add it to 'nonReservedKeywords' section.
+    keywords: [
```
Make it non reserved ?

LakshSingla replied:
It would be ideal that "CLUSTER" is non-reserved since it won't break people's query's with CLUSTER/cluster being an identifier (datasource name, column name, etc). 
But on testing it out, the test case `INSERT INTO dst SELECT * FROM foo CLUSTER BY dim1` breaks. This is because the parser is assuming it is written like `INSERT INTO dst SELECT * FROM foo AS CLUSTER ...`, since CLUSTER can be used to name stuff. Reverting the change for now.

gianm replied:
Got it. In that case let's keep it as a reserved keyword.

LakshSingla replied:
One thing that @abhishekagarwal87 suggested is to have a flag (in PlannerContext or PlannerConfig) that can be used to switch between the custom and the default parser. That would make sense in case we don't want existing queries with `cluster` identifiers to fail.

gianm replied:
IMO, we should always use our own parser. It gives us more control over our dialect, which we'll find useful for other things in the future.

I'm not sure if parser flags are a good idea or not, but if we do have them, IMO the best approach is to have "versions" of the entire parser. i.e. the flag would be `druid.sql.parser.version=2` or `druid.sql.parser.version=latest` rather than controlling specific features.

But, again, I'm not sure if we need this or not.

New keywords are added from time to time in other databases' SQL dialects. How do they typically handle this?

paul-rogers replied:
Having two parsers creates great usability and testing issues: All tests would have to be run in both. Users would need a per-query way to switch parsers, which means the poor user has to understand which is in each one, which is a hard task even for developers. Automated tools that generate Druid SQL would not know what to do.

So, better is to get the solution right within a single parser.

abhishekagarwal87 replied:
The parser flag is supposed to be at a druid cluster level. It's definitely not great but a workaround if we don't want to use a new reserved keyword. But if we do,  using `CLUSTERED BY` instead of `CLUSTER BY` will definitely make the chance of affecting queries very low so +1 to that.

LakshSingla replied:
Yep, I might have been unclear in the original comment. Its general usage is not to be overridden per query but acts as a fallback in case there is some constraint due to which these changes would break existing queries. But going through the  `PARTITIONED BY`, `CLUSTERED BY` route, it makes sense to not consider that case since maintaining and deprecating the older parser would be unnecessary overhead.

cryptoe commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlInsert.java:
```diff
@@ -0,0 +1,102 @@
+      }
+      writer.endList(frame);
+    }
+  }
```
Add a toStringMethod ?

abhishekagarwal87 replied:
thats not needed since superclass `SqlNode` already overrides `toString()` as beloe
```
public String toString() {
    return toSqlString(null).getSql();
  }
```
`toSqlString` will eventually call unparse method implemented here.

cryptoe commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -765,13 +785,53 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
+
+            if (query instanceof SqlOrderBy) {
+              SqlOrderBy sqlOrderBy = (SqlOrderBy) query;
+              // Extract the query present inside the SqlOrderBy (which is free of ORDER BY, OFFSET and FETCH clauses)
```
Maybe clarify the statement

LakshSingla replied:
Let me know if it is clearer after the update

cryptoe reviewed (commented):
LGTM post review comments.

abhishekagarwal87 commented on sql/pom.xml:
```diff
@@ -255,6 +260,140 @@
+        <artifactId>maven-dependency-plugin</artifactId>
+        <executions>
+          <execution>
+            <!-- Extract parser grammar template from Apache Calcite and put
```
can you add similar comments to other plugins that are added in this PR?

abhishekagarwal87 commented on sql/pom.xml:
```diff
@@ -255,6 +260,140 @@
         </executions>
       </plugin>
+
+      <plugin>
```
nit - are these plugins arranged in order of execution? It will be nice if they are since following the code becomes easier.

abhishekagarwal87 commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,433 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
in this file, can you add comments around custom changes that we made?

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -765,13 +785,53 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
+              }
+            }
+            // Creates a new SqlOrderBy query, which may have our CLUSTER BY overwritten
+            query = new SqlOrderBy(
```
I am wondering if the SQL layer is the right place to do this transformation or should we just leave it to the native layer to use orderBy and clusterBy together. My assumption for this transformation is that we would like the query results to be ordered on same dimensions that we want to use to arrange data in segments. Maybe it's something that QueryMaker should decide?

LakshSingla replied:
Currently, whatever you do enter in the "OrderBy" is getting converted as an ordering spec. I think the aim of the clause is to add syntactic sugar rather than add a different semantic.
(I am unfamiliar with Hive) I looked up Hive, which implements both of these clauses [stackoverflow](https://stackoverflow.com/questions/13715044/hive-cluster-by-vs-order-by-vs-sort-by), and it seems that both of them try and achieve similar functionality albeit with different results and different semantics. 
But yeah then it makes sense to implement both the CLUSTER BY and ORDER BY clauses and punt the actual logic to QueryMaker or a DruidRel. 
The trouble I see implementing it is 
a) We will need to convert it to a RelNode in SqlToRelConverter, and then pass it down the query stack. (This is because `CLUSTER BY` can be a reference to a column in the statement, and we probably need to use it). I am unsure if this is possible without making changes to the SqlToRelConverter itself. 
b) A new Query of type Query.INSERT would need to be created to accommodate "ClusteringSpec" in addition to the "OrderingSpec".

gianm replied:
I think it's OK for now to continue using the native ordering specs to represent cluster by, and to disallow having ORDER BY and CLUSTER BY at the same time. Also, IMO, we should _not_ allow ORDER BY on the SELECT for an INSERT. The error could be something like "Cannot have ORDER BY on an INSERT query; use CLUSTER BY instead."

That way, we can keep discussing if we should allow ORDER BY and CLUSTER BY to be used together, and if so, what should happen. Without blocking this PR.

paul-rogers replied:
As a counter argument, it may be good to do as much semantic analysis as possible in the SQL layer. In the fullness of time, this allows us to leverage the SQL logical plan without the conversion to native. (There is some prototyping happening in that area.) The traditional approach is to do all semantic analysis in the SQL layer so that the logical plan is validated prior to the next step (in this case, prior to generating a native query.)

The one caveat, of course, is if the information is (not yet) available to the SQL layer.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -744,18 +755,28 @@ public T next()
+
+    private ParsedNodes(
+        @Nullable SqlExplain explain,
+        @Nullable SqlInsert insert,
```
Would it make sense to have this be a DruidSqlInsert, so the granularity doesn't need to be provided separately? I'm asking since my understanding is we'll _always_ get a DruidSqlInsert, even if the user doesn't provide PARTITION BY or CLUSTER BY.

LakshSingla replied:
Yep, your understanding of the code is correct. That is we always hit our DruidSqlInsert production rule, irrespective of whether we have a simple INSERT or the one containing clauses. But just to ensure that I wasn't breaking anything that already existed while prototyping (and this overshadowing of the custom rule is not seen in the code), I am simply returning the SqlInsert statement in case both partition and cluster clauses are empty. 

Since replacing SqlInsert with DruidSqlInsert seems to be a better approach, I think we can remove separate handling of SqlInsert and DruidSqlInsert (and make the corresponding changes in the production rule)

LakshSingla replied:
Update: We might not be able to do it cleanly due to SqlExplain. Explained more in the related comment.

gianm replied:
I see. But we'll need EXPLAIN on INSERT to work eventually, so we'll need to clean this up at some point.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -765,13 +786,65 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
         query = insert.getSource();
+
+        // Processing to be done when the original query has either of the PARTITION BY or CLUSTER BY clause
+        if (insert instanceof DruidSqlInsert) {
```
When will it not be a DruidSqlInsert?

LakshSingla replied:
Updated with the context in the previous comment. Will make the necessary changes to clean up the if-else mess.

LakshSingla replied:
On removing the if-else statements, the test case like: `EXPLAIN PLAN FOR INSERT INTO dst SELECT * FROM EXTERN(...)` was failing.
This is because SqlExplain(), still doesn't know about the existence of the DruidSqlInsert node, and returns a SqlInsert() on such queries. Since we don't want the queries to fail on them, it is necessary to keep the unnessecary if-else checks. (I have cleaned up some code in the ftl file though).

Once `EXPLAIN PLAN FOR` works with the modified syntax, there would no longer be any requirements for the check. (I do have a patch that does the same, however, it involves rewriting a small part of the code from the Calcite parser, which I am wary of since we wanna avoid that). I will update the patch on the comments section of PR later (it might have become obsolete due to changing of the keywords).

paul-rogers replied:
Sounds like there needs to be a parser change so that the production rule above reduces using the Druid-specific node rather than the general one. It would see that `INSERT ...` should reduce to use the Druid node. Then `EXPLAIN` should reference the `INSERT` rule and receive the Druid-specific node. Any idea why that is not happening here?

LakshSingla replied:
Hi, as mentioned in the caveats section, what this PR achieves is adding a new top-level rule that is druid's explain the syntax. The issue is that explain is another top-level rule, which doesn't reference the druid's explain rule. Therefore `EXPLAIN PLAN FOR <druid's syntax>` is a syntactical error. 

To make it work, we can create a Druid's explain rule like: 
```
+SqlNode DruidSqlExplain() :
+{
+    SqlNode stmt;
+    SqlExplainLevel detailLevel = SqlExplainLevel.EXPPLAN_ATTRIBUTES;
+    SqlExplain.Depth depth;
+    final SqlExplainFormat format;
+}
+{
+    <EXPLAIN> <PLAN>
+    [ detailLevel = ExplainDetailLevel() ]
+    depth = ExplainDepth()
+    (
+        LOOKAHEAD(2)
+        <AS> <XML> { format = SqlExplainFormat.XML; }
+    |
+        <AS> <JSON> { format = SqlExplainFormat.JSON; }
+    |
+        { format = SqlExplainFormat.TEXT; }
+    )
+    <FOR> stmt = SqlQueryOrDmlOrDruidQuery() {
+        return new SqlExplain(getPos(),
+            stmt,
+            detailLevel.symbol(SqlParserPos.ZERO),
+            depth.symbol(SqlParserPos.ZERO),
+            format.symbol(SqlParserPos.ZERO),
+            nDynamicParams);
+    }
+}
+
+
+SqlNode SqlQueryOrDmlOrDruidQuery() :
+{
+    SqlNode stmt;
+}
+{
+    (
+        stmt = DruidSqlInsert()
+    |
+        stmt = SqlQueryOrDml()
+    ) { return stmt; }
+}
```
i.e. the `SqlQueryOrDmlOrDruidQuery` clause is new thing. But this should IMO be for another PR, since this does copy syntax from the `Parser.jj`. (That's why I refrained from adding it in the first place here).

gianm replied:
Makes sense to me to do it in a follow up. We do need to do it, though, so EXPLAIN on INSERT works properly.

gianm reviewed (commented):
Thanks @LakshSingla for the very clearly written PR description! Other than the line comments, some other comments on what's in your description:

> In the newer version of Calcite, there is a default_config.fmpp which can be used to supply the boilerplate code in config.fmpp (CALCITE-4238). On migrating to newer version, that file can be used to reduce the code in config.fmpp significantly.

For this, is there some reason we can't update right now?

> The biggest caveat to the approach is that, in the Calcite's grammar: LINK we are NOT extending the INSERT production rule.

IMO, this is totally OK and is actually a good thing. It gives us control over what we want to support in INSERT. We can take advantage of this in various ways:

- We can eliminate features we don't support from the parser, like UPSERT.
- We can _add_ other features that are Druid-specific, like some way to specify which time chunks to overwrite. (INSERT OVERWRITE, perhaps.)

> The CLUSTER BY and PARTITION BY clauses are linked to the INSERT statement, as opposed to the SELECT statements used inside them. Therefore the query looks like INSERT INTO dst (SELECT * FROM foo) PARTITION BY 'day' CLUSTER BY dim1 rather than INSERT INTO dst (SELECT * FROM foo PARTITION BY 'day' CLUSTER BY dim1).

I think this is actually a good thing too. It logically makes more sense for the PARTITION and CLUSTER info to be linked to the INSERT rather than the SELECT.

So, I think this approach you went with is pretty good!

cryptoe commented on sql/src/test/java/org/apache/druid/sql/calcite/CalciteInsertDmlTest.java:
```diff
@@ -280,6 +283,242 @@ public void testInsertFromExternal()
+  {
+    // Throws a ValidationException, which gets converted to a SqlPlanningException before throwing to end user
+    try {
+      testQuery(
```
There is a new method in `BaseCalciteQueryTest#testQueryThrows` which can be used for such test cases.

LakshSingla replied:
This works after the SQL planning has been done. We want to catch the errors before that. Therefore that won't work here.

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlInsert.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class DruidSqlInsert extends SqlInsert
+{
+  // Unsure if this should be kept as is, but this allows reusing super.unparse
```
Updated. Seemed hacky at the time, thus worded it like this.

abhishekagarwal87 commented:
@gianm re calcite upgrade, that will require us to update guava version as well. We can't do that yet because Hadoop 2 still depends on guava 2.

LakshSingla commented on sql/src/main/codegen/includes/insert.ftl:
```diff
@@ -0,0 +1,65 @@
+    insertNode = SqlInsert()
+    [
+      <PARTITION> <BY>
+      partitionBy = StringLiteral()
```
This does look nicer. Is the `TimeUnit or ALL` abstraction defined somewhere in Druid's codebase which can be reused? Or should I create a new SQL-specific class that encapsulates this?

gianm replied:
"TimeUnit or ALL" isn't defined, although maybe we could use PeriodGranularity as the class, and have the parser generate a PeriodGranularity directly?

paul-rogers replied:
Per Gian (and Hive: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-InsertingvaluesintotablesfromSQL) the arguments should be expressions, not literals. Druid may have a very limited set of expressions (such as those Gian showed), but they are still expressions.

paul-rogers replied:
Also, though I personally cringe at the idea, seems some folks like using the archaic`GROUP BY 3, 2, 5` form in `SELECT` statements. Should we allow ordinals here also? I don't like encouraging their use, but if folks already use them...

gianm replied:
> Also, though I personally cringe at the idea, seems some folks like using the archaic GROUP BY 3, 2, 5 form in SELECT statements. Should we allow ordinals here also?

I like that GROUP BY syntax because it saves typing 🙂. However, I don't think it's necessary to implement it for PARTITION BY. It'll add a bunch of complexity to the implementation and I don't think it's necessary to support it for every clause.

gianm replied:
As to what goes after PARTITION BY, I'm torn as to what's best. Context: today, partitioning must either be a floor of the primary timestamp `__time`, or "all" which means everything goes in a single partition. So I think there are three options.

1. `PARTITION BY 'day'` or `PARTITION BY 'all'`. This is what's in the PR now.
2. `PARTITION BY DAY` or `PARTITION BY ALL`. I suggested this upthread. DAY and ALL are keywords, not expressions.
3. `PARTITION BY FLOOR(__TIME TO DAY)`. I suggested this in #11929. In this case the partition-by is an expression. (I can't make up my mind, I guess, so I suggested two different things.) I didn't have a suggestion for how to say "partition by all", but IMO it should be explicit and required, so people don't accidentally partition by all when they didn't mean to.

(3) is closest to what other databases do, but those other databases generally allow partitioning on things other than time-floor. For us, it'd have to be an error to specify anything other than `FLOOR(__time TO XXX)`. That's what makes me think (2) is good: why make people specify the whole expression if it can't be changed? The obvious answer is that making things explicit is good and also that lining things up with other databases is good. But, still…

Anyway. Curious what other people think.

abhishekagarwal87 replied:
I initially suggested (3) but then moved back to camp (1) / (2) because we can only partition on time.  `PARTITIONED BY DAY/ALL`  is a much easier syntax to write. It's also probably easy to give typeahead hints with (2) (console showing a list of all possible granularities as you type PARTITIONED BY).  

I, for one, don't have a strong opinion between (1) and (2).

LakshSingla replied:
If I had to pick, 2 seems like a cleaner option. 
While 1 is implemented right now, quoting and adding a literal doesn't seem pretty, given that we have a fixed range of what can go in front of a `PARTITIONED BY`.
3 seems a bit more verbose, and to me, it implies that a) I can use a column other than `__time` b) I can use an expression other than `FLOOR`. However, I am unfamiliar with the other databases which implement the syntax in this manner and therefore of the conventions when it comes to `PARTITIONED BY`, hence I am using this convention.

LakshSingla replied:
One more question regarding the query granularities: `TimeUnit()` in Calcite is defined to have the following units (from MILLISECOND to MILLENNIUM). 
```
    <MILLISECOND> { return TimeUnit.MILLISECOND; }
|   <SECOND> { return TimeUnit.SECOND; }
|   <MINUTE> { return TimeUnit.MINUTE; }
|   <HOUR> { return TimeUnit.HOUR; }
|   <DAY> { return TimeUnit.DAY; }
|   <DOW> { return TimeUnit.DOW; }
|   <DOY> { return TimeUnit.DOY; }
|   <ISODOW> { return TimeUnit.ISODOW; }
|   <ISOYEAR> { return TimeUnit.ISOYEAR; }
|   <WEEK> { return TimeUnit.WEEK; }
|   <MONTH> { return TimeUnit.MONTH; }
|   <QUARTER> { return TimeUnit.QUARTER; }
|   <YEAR> { return TimeUnit.YEAR; }
|   <EPOCH> { return TimeUnit.EPOCH; }
|   <DECADE> { return TimeUnit.DECADE; }
|   <CENTURY> { return TimeUnit.CENTURY; }
|   <MILLENNIUM> { return TimeUnit.MILLENNIUM; }
``` 
Whereas in Druid's TimeUnitRange-Period map, the granularities are defined as 
```
  private static final Map<TimeUnitRange, Period> PERIOD_MAP = ImmutableMap.<TimeUnitRange, Period>builder()
      .put(TimeUnitRange.SECOND, Period.seconds(1))
      .put(TimeUnitRange.MINUTE, Period.minutes(1))
      .put(TimeUnitRange.HOUR, Period.hours(1))
      .put(TimeUnitRange.DAY, Period.days(1))
      .put(TimeUnitRange.WEEK, Period.weeks(1))
      .put(TimeUnitRange.MONTH, Period.months(1))
      .put(TimeUnitRange.QUARTER, Period.months(3))
      .put(TimeUnitRange.YEAR, Period.years(1))
      .build();
```
i.e. there's a mismatch as to what Calcite's SQL can accept as TimeUnit, and what we can convert using TimeUnits class. 
For the ingestion's `PARTITION BY` what range of time units from the above should be accepted by the SQL?

gianm commented:
> @gianm re calcite upgrade, that will require us to update guava version as well. We can't do that yet because Hadoop 2 still depends on guava 2.

Oh, got it. Bummer!

paul-rogers commented:
Would recommend the parser extension path: using Freemarker to assemble the full parser from a Calcite base and Druid extensions. Freemarker is new to Druid, but was used to great success for this very purpose in Apache Drill.

paul-rogers commented:
Would recommend the Hive syntax; `PARTITIONED BY` and the Hive-like `CLUSTERED BY`. The verb forms are less likely to collide with user columns. It seems likely that there are Druid use cases that use the word `cluster` (and probably `partition`) in DevOps use cases.

paul-rogers commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,443 @@
+#   3. data.parser.keywords (Added "CLUSTER")
+#   4. data.parser.statementParserMethods
+#   5. data.parser.implementationFiles
+data: {
```
This is the right approach with Calcite. For reference, here's how its done in Apache Drill: https://github.com/apache/drill/blob/9cf896a69e559cad8d1cd30d98a0555973da1ada/exec/java-exec/src/main/codegen/data/Parser.tdd

paul-rogers commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,443 @@
+    # List of new keywords. Example: "DATABASES", "TABLES". If the keyword is not a reserved
+    # keyword add it to 'nonReservedKeywords' section.
+    keywords: [
+      "CLUSTER"
```
As noted in a general comment, consider following the Hive syntax: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DML#LanguageManualDML-InsertingvaluesintotablesfromSQL

That is `CLUSTERED` and `PARTITIONED` (verb forms) which are less likely to collide with columns named `cluster` or `partition`.

gianm replied:
I agree that it makes sense to use the verb forms. Since we're adding new keywords, best to choose something that is least likely to conflict with existing columns that people may have in their tables.

paul-rogers commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,443 @@
+
+freemarkerLinks: {
+  includes: includes/
+}
```
Nit: missing newline.

paul-rogers commented on sql/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,443 @@
+    includePosixOperators: false
+    includeCompoundIdentifier: true
+    includeBraces: true
+    includeAdditionalDeclarations: false
```
`identifierMaxLength: 1024` (or some other number)?

LakshSingla replied:
It seems to be defined in the [SqlParser.configBuilder](https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerFactory.java#L53) using `.setIdentifierMaxLength()` method. Since there's no such constraint right now, should this be added?

paul-rogers replied:
Don't know, just asking. Noticed it when I compared this file with the Drill one. Use your judgement.

LakshSingla replied:
Ohh, thanks for pointing that out, I missed it. The config.fmpp that I do have is directly from the Calcite version that is being used in the current version of Druid. Better not to add this config in this PR, since it might change the current behavior of the parser.

paul-rogers commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -790,5 +862,11 @@ public SqlNode getQueryNode()
+    @Nullable
+    public String getIngestionGranularity()
+    {
+      return ingestionGranularity;
```
Would recommend "semanticising" this into something structured, such as a `PERIOD`. That tells the consumer that the value has been validated and that the consumer need not do their own validation.

paul-rogers commented on sql/src/test/java/org/apache/druid/sql/calcite/CalciteInsertDmlTest.java:
```diff
@@ -280,6 +283,242 @@ public void testInsertFromExternal()
+
+    testInsertQuery()
+        .sql(
+            "INSERT INTO druid.dst SELECT __time, FLOOR(m1) as floor_m1, dim1 FROM foo PARTITION BY 'day'")
```
Per earlier; this form of `PARTITION BY` is too implicit; where is `day` coming from? We know it is `__time`, but this makes too many assumptions. Better to use the expression Gian suggested. That way, we can have `PARTITIONED BY FLOOR(__time TO DAY), "cluster", "partition",  "node"`. (Would be good to have a better name for the function; `DAY(__time)`, say, but that's another discussion.)

paul-rogers commented on sql/src/test/java/org/apache/druid/sql/calcite/CalciteInsertDmlTest.java:
```diff
@@ -280,6 +283,242 @@ public void testInsertFromExternal()
+        () ->
+            testQuery(
+                StringUtils.format(
+                    "EXPLAIN PLAN FOR INSERT INTO dst SELECT * FROM %s PARTITION BY 'day' CLUSTER BY 1",
```
Just as general practice, it is good that `CLUSTER` supports ordinals, but, for many reasons, best if users don't actually use them.

paul-rogers commented on processing/src/main/java/org/apache/druid/query/QueryContexts.java:
```diff
@@ -68,6 +68,7 @@
   public static final String ENABLE_DEBUG = "debug";
   public static final String BY_SEGMENT_KEY = "bySegment";
   public static final String BROKER_SERVICE_NAME = "brokerService";
+  public static final String INGESTION_GRANULARITY = "ingestionGranularity";
```
Maybe we need a new class to accumulate constants for SQL statements. Maybe `Statements` in one of the SQL modules?

LakshSingla commented:
Thanks everyone for reviewing the PR. The changes have been made per the newer [proposal](https://github.com/apache/druid/issues/11929#issuecomment-1026515751) for the supported clauses in `PARTITIONED BY` and making it compulsory in `INSERT` queries. Comments from the last round of reviews have also been addressed

gianm reviewed (commented):
Looking mostly good to me; a few remaining comments about unparse and also about the error messages.

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:
```diff
@@ -0,0 +1,165 @@
+      SqlNode granularitySqlNode = operandList.get(1);
+      Preconditions.checkArgument(
+          granularitySqlNode.getKind().equals(SqlKind.LITERAL),
+          "Second argument to TIME_FLOOR in PARTITIONED BY clause should be string representing a period"
```
```suggestion
          "Second argument to TIME_FLOOR in PARTITIONED BY clause must be a period e.g. 'PT1H'"
```

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:
```diff
@@ -0,0 +1,165 @@
+        period = new Period(granularityString);
+      }
+      catch (IllegalArgumentException e) {
+        throw new ParseException(StringUtils.format("Unable to create period from %s", granularitySqlNode.toString()));
```
```suggestion
        throw new ParseException(StringUtils.format("Unable to create period from %s", granularityString));
```

LakshSingla replied:
I used the other form because then I won't need to explicitly put the string in quotes. Plus it represents exactly what the user typed, not the one curated for our use.

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:
```diff
@@ -0,0 +1,165 @@
+      // granularitySqlNode.getKind().equals(SqlKind.INTERVAL_QUALIFIER)
+      Preconditions.checkArgument(
+          granularitySqlNode instanceof SqlIntervalQualifier,
+          "Second argument to the FLOOR function in PARTITIONED BY clause cannot be converted to a valid granularity"
```
```suggestion
          "Second argument to the FLOOR function in PARTITIONED BY clause is not a valid granularity. Please refer to the documentation of FLOOR function."
```

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:
```diff
@@ -0,0 +1,165 @@
+      Preconditions.checkNotNull(
+          period,
+          StringUtils.format(
+              "Unable to convert %s to valid granularity for ingestion",
```
```suggestion
              "%s is not a valid granularity for ingestion",
```

abhishekagarwal87 commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -199,6 +203,18 @@ public PlannerResult plan() throws SqlParseException, ValidationException, RelCo
+      }
+    }
+    catch (JsonProcessingException e) {
+      throw new ValidationException("Unable to serialize partition granularity.");
```
we should log the exception stack trace.

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlInsert.java:
```diff
@@ -0,0 +1,101 @@
+  {
+    super.unparse(writer, leftPrec, rightPrec);
+    writer.keyword("PARTITIONED BY");
+    writer.keyword(getPartitionedBy().toString()); // TODO: Make it cleaner by directly unparsing the SqlNode
```
Updated the method, with the proper implementation. Here are a few unparse examples after the change:
```
INSERT INTO `dst`
(SELECT *
FROM `foo`) PARTITIONED BY ALL TIME

INSERT INTO `dst`
(SELECT *
FROM `view`.`aview`) PARTITIONED BY `TIME_FLOOR`(`__TIME`, 'PT2H')

INSERT INTO `druid`.`dst`
(SELECT `__time`, FLOOR(`m1`) AS `floor_m1`, `dim1`, CEIL(`m2`)
FROM `foo`) PARTITIONED BY FLOOR(`__TIME` TO DAY) CLUSTERED BY 2 `dim1` DESC CEIL(`m2`)
```

`unparse` method is important since this is being used in `SqlNode`'s `toString` method. So it gets used while debugging to view the original node as a SQL string. (Also, it gets used in other unparse methods, so not relevant to DruidSqlInsert currently since it's a top-level node, but it might cause say Explain on a DriudSqlInsert to fail, which is going to be implemented).

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlInsert.java:
```diff
@@ -0,0 +1,101 @@
+    writer.keyword("PARTITIONED BY");
+    writer.keyword(getPartitionedBy().toString()); // TODO: Make it cleaner by directly unparsing the SqlNode
+    if (getClusteredBy() != null) {
+      writer.sep("CLUSTERED BY");
```
Fixed, oversight on my part

LakshSingla commented:
Thanks @abhishekagarwal87 @gianm for another pass at the PR. I have updated the clumsy error messages in the DruidSqlParserUtils to be more affirmative, and hide the 'parse' that happens behind the scenes. Also, `DruidSqlInsert.unparse` now works as expected, updated the comment above with some examples for the same.

Regarding `EXPLAIN PLAN FOR druidSqlInsert`, that would for sure be added in a separate PR, and shouldn't be much work once this gets merged. One of the reasons I wanted to raise a separate PR was because this patch contains a lot of boilerplate code to make the extended syntax work and I didn't want to incorporate those changes with this patch. 

I hope I have addressed the latest and the unresolved comments. PTAL at the updated PR. Thanks.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:
```diff
@@ -38,20 +39,22 @@
+      return convertSqlNodeToGranularity(sqlNode);
     }
     catch (Exception e) {
+      log.error(e, StringUtils.format("Unable to convert %s to a valid granularity.", sqlNode.toString()));
```
If this error occurs, it's due to user error on one particular query, so it shouldn't be an ERROR that appears in the server logs. That should be reserved for really important stuff that cluster operators need to worry about.

However, there should be _some_ way to get the stack trace, and we generally _don't_ want to return it directly to users of the SQL API.

So in a situation like this I'd do `log.debug` instead of `log.error`, so server admins can see the stack trace if they really need to, but won't see it by default. That balances all of the relevant desires: server admins shouldn't see "user error" stuff by default, and users shouldn't see stack traces as part of normal error messages.

LakshSingla replied:
Updated with the correct level, thanks for the explanation.

gianm reviewed (approved):
LGTM - thanks @LakshSingla!

gianm commented:
@LakshSingla Could you please take a look at the "intellij inspections" check, and the CalciteInsertDmlTest failure in the "other modules test"? The issues there look legit.

FrankChen021 commented:
@LakshSingla The INSERT-SELECT query is only supported in the tests, right? I tried to execute a INSERT-SELECT query on the lastest master branch, it said: Cannot execute INSERT queries.

When I checked the code, I found the message was thrown from `NativeQueryMakerFactory` which implements `QueryMakerFactory`. And only the class `TestQueryMakerFactory` in the tests implements the method.

LakshSingla commented:
@FrankChen021 This PR only aims to add support for the INSERT statements with the custom PARTITIONED/CLUSTERED syntax. To actually run an INSERT query, a supported QueryMaker/QueryMakerFactory is required, which is only supported in the tests (`TestQueryMakerFactory`).

FrankChen021 commented:
@LakshSingla Thanks for the clarification. I thought the INSERT query was supported, looks like my memory is wrong.
